pkg/api/soc.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"encoding/hex"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/bmtpool"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/soc"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+type socPostResponse struct {
+	Reference swarm.Address `json:"reference"`
+}
+
+func (s *server) socSigUploadHandler(w http.ResponseWriter, r *http.Request) {
+	owner, err := hex.DecodeString(mux.Vars(r)["owner"])
+	if err != nil {
+		s.Logger.Debugf("soc upload: bad owner: %v", err)
+		s.Logger.Error("soc upload: bad owner")
+		jsonhttp.BadRequest(w, "bad owner")
+		return
+	}
+	id, err := hex.DecodeString(mux.Vars(r)["id"])
+	if err != nil {
+		s.Logger.Debugf("soc upload: bad id: %v", err)
+		s.Logger.Error("soc upload: bad id")
+		jsonhttp.BadRequest(w, "bad id")
+		return
+	}
+	sig, err := hex.DecodeString(mux.Vars(r)["sig"])
+	if err != nil {
+		s.Logger.Debugf("soc upload: bad signature: %v", err)
+		s.Logger.Error("soc upload: bad signature")
+		jsonhttp.BadRequest(w, "bad signature")
+		return
+	}
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		if jsonhttp.HandleBodyReadError(err, w) {
+			return
+		}
+		s.Logger.Debugf("soc upload: read chunk data error: %v", err)
+		s.Logger.Error("soc upload: read chunk data error")
+		jsonhttp.InternalServerError(w, "cannot read chunk data")
+		return
+	}
+	if len(data) < swarm.SpanSize {
+		s.Logger.Debugf("soc upload: chunk data too short")
+		s.Logger.Error("soc upload: chunk data")
+		jsonhttp.BadRequest(w, "short chunk data")
+		return
+	}
+
+	ch, err := chunk(data)
+	if err != nil {
+		s.Logger.Debugf("soc upload: create content addressed chunk: %v", err)
+		s.Logger.Error("soc upload: chunk data error")
+		jsonhttp.BadRequest(w, "chunk data error")
+		return
+	}
+
+	chunk, err := soc.NewSignedChunk(id, ch, owner, sig)
+	if err != nil {
+		s.Logger.Debugf("soc upload: read chunk data error: %v", err)
+		s.Logger.Error("soc upload: read chunk data error")
+		jsonhttp.InternalServerError(w, "cannot read chunk data")
+		return
+	}
+
+	if !soc.Valid(chunk) {
+		s.Logger.Debugf("soc upload: invalid chunk: %v", err)
+		s.Logger.Error("soc upload: invalid chunk")
+		jsonhttp.BadRequest(w, "invalid chunk")
+		return
+
+	}
+	ctx := r.Context()
+
+	_, err = s.Storer.Put(ctx, requestModePut(r), chunk)
```

Comment:
why request mode? it should be `uploadModePut`

Reply:
gets the `ModePut` for this request. therefore `requestModePut`

Reply:
no no, `requestModePut` is for chunks that arrive as response to retrieve requests.  this must be `uploadModePut`